### DISCUSSION TITLE
Implementing useState similar to React to achieve stateful functional components

### DISCUSSION THREAD

**Discussion by B4PT0R** [reflex-dev/reflex#3225] [Ideas]
Hello,

Just discovered Reflex recently. Amazing piece of work here.

I've been lurking at the source code and played around with the framework to get to know its internal workings better.

Here is what I came up with : 

```python
import reflex as rx
import uuid

def use_state(default,vartype=None):
    """
    Creates a state with a single var 'value' set to default and returns the corresponding state var and setter
    """
    vartype=vartype or type(default)
    attributes={
        'value':default,
        '__annotations__':{'value':vartype}
    }
    cls_name="State_"+str(uuid.uuid4())
    state=type(cls_name,(rx.State,),attributes)
    state_var=state.value
    state_setter=state.set_value
    return state_var,state_setter

#Usage:

def button(initial_count:int=0) -> rx.Component:
    count,set_count=use_state(initial_count)
    return rx.button(count,on_click=set_count(count+1))

@rx.page()
def index() -> rx.Component:
    return rx.box(
        button(),
        button(10)
    )

app = rx.App()
```

Which successfully produces two stateful buttons with independent states.
The state is purely internal to the button and can't be referenced from outside.
Idk, maybe this has some relevance for some use cases where you just want to quickly use a state in a purely functional component.

**Comment by benedikt-bartscher** [comment]
@B4PT0R there is `rx.ComponentState` for stateful components, docs: https://reflex.dev/docs/substates/component-state/

**Reply by B4PT0R** [reply]
@benedikt-bartscher Yes, I know. I looked at the source code of ComponentState to figure out how to implement this.
It's mostly an alternative to quickly spawn a state variable/setter into a functional component without having to declare a full ComponentState object. Akin to the difference in React between class-based and functional components.
I admit it won't change the face of the world, I was just playing with the library.

**Comment by masenf** [comment]
Some prior art:

https://github.com/reflex-dev/reflex/blob/9c7dbdbc728c34611fdd9328a8daa942c2c68cb8/reflex/experimental/hooks.py#L80-L93

https://github.com/masenf/reflex-pomodoro/blob/main/reflex_pomodoro/components/use_state.py

https://github.com/reflex-dev/reflex/issues/2917#issuecomment-2019346612 (similar to above, but with a small working example)

> The team has been discussing the exposure of client-side-only state (i.e. formalizing the useState function I have defined above), but we haven't decided if it will actually help, or make things more confusing... the requirement to wrap components in an IntegralFragment to allow multiple components to share the client side state is also a bit weird. [[1]](https://github.com/reflex-dev/reflex/issues/2917#issuecomment-2025949546)

(Edit: oops, misread this and thought the code was going for browser-side state. At any rate, leaving it for potential inspo)
